[move-prover] transform a loop into a DAG

mengxu-fb opened the pull request:
This commit breaks the back-edge of a loop and place the induction case
checking of the loop invariant in a newly created block.

The general procedure works as following (assuming the loop invariant expression is L):

- Collect loop targets (T) - the set of variables that can be possibly mutated during the execution of the loop.
  - Loop targets are categorized into three kinds:
    - value targets: the temporary (e.g. `$t1`) is a Value type and can simply be `havoc`-ed.
    - mutation targets that changes values only, e.g., a `WriteBack` to a `Reference` (or a `WriteRef`) to `$t2`, we know that only the value of `$t2` is changed, the location and path of `$t2` do not change.
    - mutation targets that changes both its pointer and values, e.g., re-assign a `&mut` to another location.
- At the beginning of the loop header (identified by the label bytecode), insert the following statements:
    - assert L;
    - havoc T;
        - A value target will be translated to `havoc $t_n`
        - A mutation target that changes its value only will be translated to `call $t_n = $HavocMutation($t_n)`
        - A mutation target that changes its pointer and value will be translated to `havoc $t_n`
    - assume L;
- Create a new dummy block (say, block X) with only the following statements
    - assert L;
    - stop;
- For each backedge in this loop:
    - In the source block of the backedge, replace the last statement (must be a jump or branch) with the new label of X

To support the algorithm above, data structures used in Loop calculation and analysis are changed as well and invariants (assumptions that enables the loop transformation algorithm) is checked. In particular,

  -  The `compute_reducible()` function now returns a vector of `NaturalLoop` which encapsulates the information on `loop_header`, `loop_body`, and the `loop_latch` (latch -> header is the backedge).
  - It also brings the following checks explicit (i.e., the prover panics)
    - if the control-flow graph is not reducible
    - if the loop header block is a dummy block or the first bytecode in the header block is not a label

As a result of the transformation, we can attach different error messages to the `assert`s for the base case and induction case. See the test case in `move-prover/tests/sources/functional/loops.move` and the associated exp file for an example.

A verification in `Diem.move`, function `remove_preburn_from_queue` is re-enabled and verifies successfully after the loop transformation.

## Motivation

- simplifies encoding of loops
- separation of loop invariant base case proving and induction case proving

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

`cargo xtest`, the related test cases are updated.

shazqadeer commented:
Are we planning to land this before the upcoming release?

mengxu-fb commented:
No, that is not the intention. My understanding is that we will land it at least after the release.

shazqadeer commented:
The PR description says that a separate dummy block is generated for each back edge.  I suppose you can use the same dummy block for all back edges of a loop.

mengxu-fb commented:
My bad, the implementation is actually one dummy block for  the whole loop and stitch all backedges to that dummy block. I'll update the description.

mengxu-fb commented:
More work items for this PR based on the discussion with Shaz:

1. Fix the wrong assumption that two loops cannot share the same header.
2. Write an analysis to collect mutable references that do not have its location/path changed in loop body and re-establish the assumptions after havoc.

The following items also come into discussion but may be handled in separate PRs after this PR lands:

3. Allow the use of `old(.)` in loop invariants.
4. Have separate error messages differentiating the violation of loop invariants upon entry and upon induction.

Marking the PR under draft until item 1, 2 are fixed.

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

mengxu-fb commented:
Major changes after the initial version

1. Fixed the wrong assumption that two loops cannot share the same header.
  - Now, if two loops share the same header (e.g., due to two `continue` statements in the loop body), we construct a fat-loop to capture both loops.
2. Added an analysis to collect mutable references that do not have its location/path changed in loop body and re-establish the assumptions after havoc.
  - The rules are captured in `loop_analysis.rs`, function `targets()`.
3. Ported a few representative loop tests to move-prover, including
  - nested loops
  - loops with two back edges
  - loops with memory operations in it

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

wrwg reviewed (changes requested):
Nice!

mengxu-fb commented on language/move-prover/boogie-backend-v2/src/bytecode_translator.rs:
```diff
@@ -1030,25 +1030,42 @@ impl<'env> ModuleTranslator<'env> {
                     }
+                    Stop => {
+                        emitln!(self.writer, "assume false;");
+                        // a return statement terminates any execution trace that reaches it
```
It should be the combination of both. I'll update it.

mengxu-fb commented on language/move-prover/bytecode/src/loop_analysis.rs:
```diff
@@ -54,43 +91,142 @@ impl FunctionTargetProcessor for LoopAnalysisProcessor {
-        data: &FunctionData,
-    ) -> BTreeMap<Label, BTreeSet<usize>> {
+    fn build_loop_annotation(func_env: &FunctionEnv<'_>, data: &FunctionData) -> LoopAnnotation {
+        // build for natural loops
```
Yes, this is a large one. I think I can split it into at least three sub functions.

mengxu-fb commented on language/move-prover/bytecode/src/loop_analysis.rs:
```diff
@@ -107,63 +243,156 @@ impl LoopAnalysisProcessor {
+    fn targets(
+        func_target: &FunctionTarget<'_>,
+        bytecode: &Bytecode,
+    ) -> (Vec<TempIndex>, Vec<(TempIndex, bool)>) {
```
I see, that makes sense. I'll see whether they can be unified.

mengxu-fb replied:
Unified with `bc.modifies()`, and found a couple of missing cases already, thank you for the suggestion!

Now the `Bytecode::modifies()` function will return `(Vec<TempIndex>, Vec<(TempIndex, bool)>)`,
- the first Vec contains the Values that are modified
- the second Vec contains the Mutations that are modified as well as how they are modified (i.e., whether the whole reference is modified or just the underlying value is modified)

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

shazqadeer commented on language/move-prover/boogie-backend-v2/src/prelude.bpl:
```diff
@@ -711,6 +711,13 @@ procedure {:inline 1} $HavocMutation(m: $Mutation) returns (m': $Mutation) {
 }
 
+procedure {:inline 1} $HavocMutationAll(m: $Mutation) returns (m': $Mutation) {
+  var l': $Location;
```
You can also just get the same effect by replacing

call p := $HavocMutationAll(p)

with 

havoc p.

You are really just havocing the variable entirely in this case also.

mengxu-fb replied:
Right! In this way we can also remove the `$HavocMutationAll` helper.

mengxu-fb commented on language/move-prover/boogie-backend-v2/src/bytecode_translator.rs:
```diff
@@ -1030,25 +1030,42 @@ impl<'env> ModuleTranslator<'env> {
                         );
                     }
-                    Havoc => {
+                    HavocVal => {
```
Unified `HavocVal` and `HavocMut` into `Havoc(HavocKind)`.

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

mengxu-fb commented on language/move-prover/bytecode/src/loop_analysis.rs:
```diff
@@ -54,43 +91,142 @@ impl FunctionTargetProcessor for LoopAnalysisProcessor {
+            .collect();
+
+        for (label, loop_info) in &loop_annotation.fat_loops {
+            let checker_label = invariant_checker_labels.get(label).unwrap();
```
My personal preference is to keep them in the same function as the code represents the whole loop-to-DAG transformation algorithm, and it might hurt readability if we force some blocks into a helper function.

This `transform` is a bit different from the other large function `build_loop_annotation`. The latter can indeed be sliced into three helper functions, but it is hard to do so for `transform`.

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

mengxu-fb commented on language/move-prover/bytecode/src/loop_analysis.rs:
```diff
@@ -107,63 +243,156 @@ impl LoopAnalysisProcessor {
     }
 
-    fn targets(bytecode: &Bytecode) -> Vec<usize> {
+    fn targets(
```
Unified with `bc.modifies()` and extended the comments on that function.

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

mengxu-fb commented on language/move-model/src/model.rs:
```diff
@@ -1411,6 +1411,11 @@ impl<'env> ModuleEnv<'env> {
     }
 
+    /// Gets the source map for the module
+    pub fn get_source_map(&'env self) -> &'env SourceMap<MoveIrLoc> {
```
Function removed, disassembling is localized in move-model.

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

mengxu-fb commented:
@wrwg comments addressed and suggestions incorporated, all in the last 6 commits.

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

mengxu-fb commented:
Rebased to master to resolve the conflict of automated doc file `Diem.md`, nothing changed

libra-action commented:
**release binaries dependency change summary**:
```
target packages:
  A disassembler 0.1.0 (workspace, path 'language/tools/disassembler')
    * features: default
  A move-coverage 0.1.0 (workspace, path 'language/tools/move-coverage')
    * features: default
  A colored 2.0.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"release-summary"} -->

mengxu-fb commented:
/land

bors-libra commented:
:broken_heart: Test Failed - [ci-test-success](https://github.com/diem/diem/runs/2249915681)

mengxu-fb commented:
/land

bors-libra commented:
:broken_heart: Test Failed - [ci-test-success](https://github.com/diem/diem/runs/2250644528)

mengxu-fb commented:
Rebased to latest `main` to see if the land-blocking tests are passing.

mengxu-fb commented:
/land
